### PR TITLE
Update deepchem.feat.graph_data.GraphData

### LINKED ISSUE #2824: Updating deepchem.feat.graph_data.GraphData
## 🚀 Feature

DeepChem currently uses `dc.feat.graph_data.GraphData` for handling graphs. I think it will be more apt for `GraphData` to be under the `dc.data`, thereby making them a fundamental data type supported by DeepChem.

### Motivation
- Graph based data types seems to be a very common kind of data type used in scientific applications.
- We can add more support for handling graphs and performing operations on them - like saving, loading, converting it to other formats etc

### Additional Context on transition

For doing the transition, we can probably begin with createing `dc.data.GraphDataset` class like `dc.data.NumpyDataset` which can take in node features and edge features - just like `dc.feat.graph_data.GraphData` but under `dc.data`. Next, we can add a depreciation note and update code in repository to use dc.data.GraphDataset. Later on, we can add more support on handling (like saving, retrieving) graphs.

Some features which will be nice to see (feel free to add yours):
- Conversion of pytorch-geometric, networkx and other graph forms to dc.data.GraphDataset and vice-versa
- Utilities to compute graph properties like centrality measures, etc

### PR DESCRIPTION
**Pull request by neerajanand-coder**
# Pull Request Template

## Description

Fix #2824 

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [X] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by arunppsg** [comment]
Hey, let me know when this ready for review.

**Comment by neerajanand-coder** [comment]
Hi @arunppsg 
I am unable to figure out how to add the second feature (storing global graph level features). Could you please help me so that I make this pr ready to merge.

**Comment by arunppsg** [comment]
I would suggest adding a kwargs argument to `__init__` call of `GraphData` and then we can sett values for those attributes in kwargs. This way user can add any features they want into graph data. An example usage:
```
class Foo(object):
    
    def __init__(self, **kwargs):
        for key, value in kwargs.items():
            setattr(self, key, value)
```

File: deepchem/feat/graph_data.py [hunk 1]
```diff
@@ -90,7 +90,26 @@ def __init__(
     self.num_edges = edge_index.shape[1]
     if self.edge_features is not None:
       self.num_edge_features = self.edge_features.shape[1]
+  
+  def __repr__(self):
+    """Utility to print the graph in more useful way.
+
```
Comment (arunppsg) [inline deepchem/feat/graph_data.py:96]: Probably we don't need an example here. Instead you can combine it with this example https://github.com/deepchem/deepchem/blob/059ffa00f3e22e23447f99e11d911e0e6e8d5c5d/deepchem/feat/graph_data.py#L35

File: deepchem/feat/graph_data.py [hunk 2]
```diff
@@ -90,7 +90,26 @@ def __init__(
...
+    >>> graph = GraphData(node_features=node_features, edge_index=edge_index)
+    >>> print(graph)
 
+    Returns
+    -------
+        GraphData(node_features=[5, 10], edge_index=[2, 5])
+     """
+    node_features_str = str(list(node_features.shape))
+    edge_index_str = str(list(edge_index.shape))
+    return  "GraphData(node_features="+node_features_str+", edge_index="+edge_index_str+")"
```
Comment (arunppsg) [inline deepchem/feat/graph_data.py:111]: You need to use string formatting here before returning here. It improves readability.

File: deepchem/feat/graph_data.py [hunk 3]
```diff
@@ -90,7 +90,26 @@ def __init__(
...
+    >>> edge_index = np.array([[0, 1, 2, 3, 4], [1, 2, 3, 4, 0]], dtype=np.int64)
+    >>> graph = GraphData(node_features=node_features, edge_index=edge_index)
+    >>> print(graph)
 
+    Returns
+    -------
+        GraphData(node_features=[5, 10], edge_index=[2, 5])
+     """
+    node_features_str = str(list(node_features.shape))
+    edge_index_str = str(list(edge_index.shape))
```
Comment (arunppsg) [inline deepchem/feat/graph_data.py:110]: Maybe also add edge_features shape?

File: deepchem/feat/graph_data.py [hunk 4]
```diff
@@ -90,6 +91,20 @@ def __init__(
     self.num_edges = edge_index.shape[1]
     if self.edge_features is not None:
       self.num_edge_features = self.edge_features.shape[1]
+    for key, value in kwargs.items():
+      setattr(self, key, value)
+
+  def __repr__(self):
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:97]: We should have a short docstring here explaining what information it displays

**Review by rbharath — COMMENTED** [review]
Looks good! One brief comment here about docstring

**Comment by arunppsg** [comment]
Closing this since the commits went some what our of order. Put up a new one #2860

**Pull request closed without merging**